File: app/src/main/java/de/bentolor/ahamanager/AhaFragment.java
package de.bentolor.ahamanager;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.NavUtils;
import android.text.Editable;
import android.text.TextWatcher;
import android.text.format.DateFormat;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.CompoundButton.OnCheckedChangeListener;
import android.widget.EditText;

import java.util.Date;
import java.util.UUID;

public class AhaFragment extends Fragment {
    public static final String EXTRA_AHA_ID = "de.bentolor.ahamoment.aha_id";
    private static final String DIALOG_DATE = "date";
    private static final String DIALOG_TIME = "time";
    private static final int REQUEST_DATE = 0;
    private static final int REQUEST_TIME = 1;

    private Aha mAha;
    private Button mDateButton;
    private Button mTimeButton;

    public static AhaFragment newInstance(UUID ahaId) {
        Bundle args = new Bundle();
        args.putSerializable(EXTRA_AHA_ID, ahaId);

        AhaFragment fragment = new AhaFragment();
        fragment.setArguments(args);

        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        UUID ahaID = (UUID) getArguments().getSerializable(EXTRA_AHA_ID);

        mAha = AhaLab.get(getActivity()).getAha(ahaID);

        setHasOptionsMenu(true);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_aha, container, false);

        if (NavUtils.getParentActivityName(getActivity()) != null) {
            getActivity().getActionBar().setDisplayHomeAsUpEnabled(true);
        }

        EditText mTitleField = (EditText) v.findViewById(R.id.aha_title);
        mTitleField.setText(mAha.getTitle());
        mTitleField.addTextChangedListener(new TextWatcher() {

            @Override
            public void onTextChanged(CharSequence c, int start, int before, int count) {
                mAha.setTitle(c.toString());
            }

            @Override
            public void beforeTextChanged(CharSequence c, int start, int count, int after) {
            }

            @Override
            public void afterTextChanged(Editable c) {
            }
        });

        mDateButton = (Button) v.findViewById(R.id.aha_date);
        mDateButton.setOnClickListener(new View.OnClickListener() {

            @Override
            public void onClick(View dateButton) {
                FragmentManager fm = getActivity().getSupportFragmentManager();
                DatePickerFragment dialog = DatePickerFragment.newInstance(mAha.getDate());
                dialog.setTargetFragment(AhaFragment.this, REQUEST_DATE);
                dialog.show(fm, DIALOG_DATE);
            }
        });

        mTimeButton = (Button) v.findViewById(R.id.aha_time);
        mTimeButton.setOnClickListener(new View.OnClickListener() {

            @Override
            public void onClick(View timeButton) {
                FragmentManager fm = getActivity().getSupportFragmentManager();
                TimePickerFragment dialog = TimePickerFragment.newInstance(mAha.getDate());
                dialog.setTargetFragment(AhaFragment.this, REQUEST_TIME);
                dialog.show(fm, DIALOG_TIME);
            }
        });

        updateDate();

        CheckBox usefulBox = (CheckBox) v.findViewById(R.id.aha_useful);
        usefulBox.setChecked(mAha.isUseful());
        usefulBox.setOnCheckedChangeListener(new OnCheckedChangeListener() {
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                mAha.setUseful(isChecked);
            }
        });

        return v;
    }

    private void updateDate() {
        Date d = mAha.getDate();
        CharSequence c = DateFormat.format("EEEE, MMM dd, yyyy", d);
        CharSequence t = DateFormat.format("h:mm a", d);
        mDateButton.setText(c);
        mTimeButton.setText(t);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (resultCode != Activity.RESULT_OK) return;
        if (requestCode == REQUEST_DATE) {
            Date date = (Date) data.getSerializableExtra(DatePickerFragment.EXTRA_DATE);
            mAha.setDate(date);
            updateDate();
        } else if (requestCode == REQUEST_TIME) {
            Date date = (Date) data.getSerializableExtra(TimePickerFragment.EXTRA_TIME);
            mAha.setDate(date);
            updateDate();
        }
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.fragment_aha, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                if (NavUtils.getParentActivityName(getActivity()) != null) {
                    NavUtils.navigateUpFromSameTask(getActivity());
                }
                return true;
            case R.id.menu_item_delete_crime:
                AhaLab ahaLab = AhaLab.get(getActivity());
                ahaLab.deleteAha(mAha);
                ahaLab.saveAhas();
                getActivity().finish();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public void onPause() {
        super.onPause();
        AhaLab.get(getActivity()).saveAhas();
    }
}

File: app/src/main/kotlin/de/bentolor/katamanager/AhaFragment.kt
package de.bentolor.katamanager

import android.R.id
import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.app.NavUtils
import android.support.v4.app.NavUtils.getParentActivityName
import android.support.v4.app.NavUtils.navigateUpFromSameTask
import android.text.Editable
import android.text.TextWatcher
import android.text.format.DateFormat
import android.text.format.DateFormat.format
import android.view.*
import de.bentolor.ahamanager.R
import kotlinx.android.synthetic.main.fragment_aha.*
import java.util.*

class AhaFragment : Fragment() {

    private var mAha: Aha = Aha()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val ahaID = arguments.getSerializable(EXTRA_AHA_ID) as UUID

        mAha = AhaLab.get(activity).getAha(ahaID) ?: Aha()

        setHasOptionsMenu(true)
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val v = inflater!!.inflate(R.layout.fragment_aha, container, false)

        if (getParentActivityName(activity) != null) {
            activity.actionBar!!.setDisplayHomeAsUpEnabled(true)
        }

        return v
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        aha_title.setText(mAha.title)
        aha_title.addTextChangedListener(object : TextWatcher {
            override fun onTextChanged(c: CharSequence, start: Int, before: Int, count: Int) {
                mAha.title = c.toString()
            }

            override fun beforeTextChanged(c: CharSequence, start: Int, count: Int, after: Int) {
            }

            override fun afterTextChanged(c: Editable) {
            }
        })

        aha_date.setOnClickListener {
            val dialog = DatePickerFragment.newInstance(mAha.date)
            dialog.setTargetFragment(this@AhaFragment, REQUEST_DATE)
            dialog.show(activity.supportFragmentManager, DIALOG_DATE)
        }

        aha_time.setOnClickListener {
            val dialog = TimePickerFragment.newInstance(mAha.date)
            dialog.setTargetFragment(this@AhaFragment, REQUEST_TIME)
            dialog.show(activity.supportFragmentManager, DIALOG_TIME)
        }

        updateDate()

        aha_useful.isChecked = mAha.isUseful
        aha_useful.setOnCheckedChangeListener { buttonView, isChecked -> mAha.isUseful = isChecked }

    }

    private fun updateDate() {
        aha_date.text = format("EEEE, MMM dd, yyyy", mAha.date)
        aha_time.text = format("h:mm a", mAha.date)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (resultCode != Activity.RESULT_OK) return
        if (requestCode == REQUEST_DATE) {
            val date = data!!.getSerializableExtra(DatePickerFragment.EXTRA_DATE) as Date
            mAha.date = date
            updateDate()
        } else if (requestCode == REQUEST_TIME) {
            val date = data!!.getSerializableExtra(TimePickerFragment.EXTRA_TIME) as Date
            mAha.date = date
            updateDate()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater!!.inflate(R.menu.fragment_aha, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        when (item!!.itemId) {
            R.id.home -> {
                if (getParentActivityName(activity) != null) {
                    navigateUpFromSameTask(activity)
                }
                return true
            }
            R.id.menu_item_delete_crime -> {
                with(AhaLab[activity]) {
                    deleteAha(mAha)
                    saveAhas()
                }
                activity.finish()
                return true
            }
            else -> return super.onOptionsItemSelected(item)
        }
    }

    override fun onPause() {
        super.onPause()
        AhaLab[activity].saveAhas()
    }

    companion object {
        val EXTRA_AHA_ID = "de.bentolor.ahamoment.aha_id"
        private val DIALOG_DATE = "date"
        private val DIALOG_TIME = "time"
        private val REQUEST_DATE = 0
        private val REQUEST_TIME = 1

        fun newInstance(ahaId: UUID): AhaFragment {
            val args = Bundle()
            args.putSerializable(EXTRA_AHA_ID, ahaId)

            val fragment = AhaFragment()
            fragment.arguments = args

            return fragment
        }
    }
}
